### ISSUE TITLE
Unable to run again after crashing

### ISSUE THREAD

**Issue by hamid13r** [warpem/miss-alignment#105]
I am trying to test miss-alignment on our K3 data and I keep getting the std=0 error. While I am changing parameters to see how to get around that something weird is happening. 

I get this error
```
[rank1]:[E513 12:38:14.493976596 ProcessGroupNCCL.cpp:1870] [PG ID 0 PG GUID 0(default_pg) Rank 1] ProcessGroupNCCL's watchdog got stuck for 480 seconds without making progress in monitoring enqueued collectives. This typically indicates a NCCL/CUDA API (e.g., CudaEventDestroy) hang blocking the watchdog, and could be triggered by another thread holding the GIL inside a CUDA api (for example, CudaEventDestroy), or other deadlock-prone behaviors.If you suspect the watchdog is not actually stuck and a longer timeout would help, you can either increase the timeout (TORCH_NCCL_HEARTBEAT_TIMEOUT_SEC) to a larger value or disable the heartbeat monitor (TORCH_NCCL_ENABLE_MONITORING=0).If either of aforementioned helps, feel free to file an issue to PyTorch about the short timeout or false positive abort; otherwise, please attempt to debug the hang.
[rank1]:[E513 12:38:14.494343247 ProcessGroupNCCL.cpp:1589] [PG ID 0 PG GUID 0(default_pg) Rank 1] ProcessGroupNCCL preparing to dump debug info. Include stack trace: 1
[rank1]:[F513 12:46:14.495035780 ProcessGroupNCCL.cpp:1614] [PG ID 0 PG GUID 0(default_pg) Rank 1] [PG ID 0 PG GUID 0(default_pg) Rank 1] Terminating the process after attempting to dump debug info, due to ProcessGroupNCCL watchdog hang.
[rank: 1] Child process with PID 334187 terminated with code -6. Forcefully terminating all other processes to avoid zombies 🧟
```

and it goes away if I reboot the machine. (sometimes it runs again)

Is this something that can be solved?

**Comment by McHaillet** [comment]
Hmm, sounds like something to do with hanging processes. When the program crashes, can you see with htop or nvidia-smi whether there are still hanging (zombie) processes from miss-alignment?

**Comment by hamid13r** [comment]
I don't see anything in nvidia-smi, sometimes htop shows some hanging processes which I clear with 
```bash
pkill -9 -f miss-alignment
```
but I still get the same error. It is not very consistent but it does go away after rebooting.

**Comment by hamid13r** [comment]
I have an update on this! This `ProcessGroupNCCL's watchdog got stuck for 480 seconds without making progress` kept happening and I used claude to diagnose the issue. It concluded that "P2P is broken on this host" and gave me a workaround. I am not really sure what this means but here is the summary I got in case it would be important for debugging/development:

<details>
<summary> Claude's summary, click to expand. </summary>
# miss-alignment multi-GPU hang — root cause: broken PCIe GPU P2P (not a miss-alignment bug)

**Summary:** `miss-alignment train` with `--training-devices 0,1` hangs forever right after the
reconstruction pool is created. The cause is outside miss-alignment: NCCL peer-to-peer transport
is broken on this host, so the first DDP collective never completes. Setting
`NCCL_P2P_DISABLE=1` fixes it completely and training runs normally.

## Environment

| | |
|---|---|
| miss_alignment | 0.2.0 |
| torch | 2.8.0+cu128 (CUDA 12.8) |
| NCCL | 2.27.3 |
| Python | 3.11 (micromamba env `miss-alignment`) |
| GPUs | 4x RTX A5000 (GA102GL, `10de:2231`) |
| Driver | 535.274.02 |
| Host | AMD platform, IOMMU enabled (`ivhd0..3`), no NVLink |

## Command that hangs

```
CUDA_VISIBLE_DEVICES=0,1,2,3 OMP_NUM_THREADS=1 MKL_NUM_THREADS=1 \
miss-alignment train --config-file config.yaml --training-devices 0,1 \
  --reconstruction-devices 2,2,2,3,3,3 --dataloaders-per-trainer 2 \
  --start-at-iteration 1 --prepare-stacks 15
```

Last output, then silence indefinitely (no traceback; killed before the default 1800 s NCCL
watchdog could fire):

```
Iteration 2/8 - Alignment: anchoring
Created pool directory: /tmp/recon_pool_c5a48e432f0b
Pool configuration: 6 workers -> 4 partitions (250 files each)
```

## Evidence the hang is upstream of the data pipeline

Filesystem timestamps from the hung run:

```
pool dir created           23:38:39.756
1000 pool files written    23:38:44.20 -> 23:38:48.93   (4.7 s)
files per partition        250 / 250 / 250 / 250        (== partition_size, i.e. exactly full)
files consumed             0
```

Interpretation:

1. The 6 reconstruction workers are **healthy** — they filled the pool and then correctly
   blocked on backpressure in `_reconstruction_worker.py` ("all partitions full" sleep loop).
2. **Zero files were ever consumed.** `ReconstructionPoolDataset.__getitem__` deletes each file
   after reading, and DataLoader workers begin prefetching as soon as they are forked. Nothing
   was deleted, so the DataLoader workers were **never forked** — execution never reached
   `train_dataloader()`.
3. That also **rules out `torch.compile`**, which only triggers on the first forward pass.
4. No new `models/lightning_logs/version_N` directory was created.

So the hang sits between `datamodule.__enter__()` and the first batch — i.e. in DDP /
process-group setup.

## Minimal reproducer (no miss-alignment involved)

Standalone script: `mp.spawn` 2 ranks -> `init_process_group("nccl")` -> `all_reduce` -> DDP
fwd/bwd, with a 60 s timeout.

```
$ CUDA_VISIBLE_DEVICES=0,1,2,3 python -u nccl_p2p_test.py
[rank 0] pre init_process_group
[rank 1] pre init_process_group
[rank 1] process group READY        <-- rendezvous SUCCEEDS
[rank 0] process group READY
NCCL version 2.27.3+cuda12.9
                                    <-- first collective never completes
[rank1] Watchdog caught collective operation timeout:
        WorkNCCL(SeqNum=1, OpType=ALLREDUCE, NumelIn=1024, Timeout(ms)=60000)
        ran for 60084 milliseconds before timing out.
[rank0] ... last enqueued work: 1, last completed work: -1
-> ProcessExitedException: process 0 terminated with signal SIGABRT
```

Rendezvous completes on both ranks; the **first `all_reduce` of 1024 floats** hangs. Same script
with P2P disabled:

```
$ NCCL_P2P_DISABLE=1 python -u nccl_p2p_test.py
[rank 0] all_reduce OK -> 3.0 (expect 3.0)
[rank 1] all_reduce OK -> 3.0 (expect 3.0)
[rank 0] DDP fwd/bwd OK
[rank 1] DDP fwd/bwd OK
ALL NCCL TESTS PASSED
```

## Why P2P is broken on this host

```
$ nvidia-smi topo -m
      GPU0  GPU1  GPU2  GPU3
GPU0   X    PHB   SYS   SYS
GPU1  PHB    X    SYS   SYS

$ cat /proc/cmdline
BOOT_IMAGE=... ro quiet splash vt.handoff=7        # no iommu=pt

$ # each GPU isolated in its own IOMMU group -> ACS active
group 51 : 0000:21:00.0    group 36 : 0000:41:00.0
group 52 : 0000:22:00.0    group 37 : 0000:43:00.0

$ lspci -tv
+-01.1-[21]-- GPU0     +-01.1-[41-42]-- GPU2
+-03.1-[22]-- GPU1     +-03.1-[43]----- GPU3      # separate root ports
```

The GPU pairs sit behind separate PCIe root ports, so peer traffic must traverse the host bridge
(`PHB`) and be translated by the AMD IOMMU, which is in full translation mode. This is the
configuration NVIDIA documents as requiring IOMMU off / passthrough for GPU P2P. Host-side fix
(untested, pending reboot): `iommu=pt` on the kernel cmdline, or disable ACS in BIOS.

## Confirmed working workaround

```
NCCL_P2P_DISABLE=1 miss-alignment train ...
```

Training now runs normally: `Training: 4500/15000 [26:22<1:01:32, 2.84step/s, loss=0.0238,
aligned=4.019, misaligned=11.543, epoch=8/30]`, pool draining as expected (230-244 files per
partition, below the 250 cap).

Performance cost is negligible for this model: 295,825 parameters => ~1.18 MB gradient allreduce
per step, ~3.4 MB/s at 2.84 step/s.

## Suggestions for miss-alignment

Not bugs, but they made a 15-minute problem take much longer to find:

1. **Preflight the process group.** Before spawning the reconstruction pool, run a tiny
   `all_reduce` with a short timeout (~60 s) when `world_size > 1`. It would turn this silent
   hang into an immediate, actionable error — and it would avoid spending GPU time filling a
   1000-file pool that then can't be consumed.
2. **Consider a shorter default DDP timeout,** or document it. 1800 s of total silence is
   indistinguishable from a deadlock.
3. **Lightning's DDP banners are silenced** in `utils.configure_logging()`
   (`lightning.pytorch`/`lightning.fabric` -> WARNING). That hides
   `Initializing distributed: GLOBAL_RANK ...`, which is exactly the breadcrumb needed to see
   that the hang was in process-group init. Maybe keep it at INFO, or log one explicit
   "initializing process group (world_size=N)" line.
4. **stdout is block-buffered when redirected to a file,** so `print()` output flushes late while
   tqdm (stderr) appears immediately. Ordering in a redirected log is therefore misleading — the
   run looked frozen at the pool line when it had actually moved past it. `flush=True` on the
   phase prints (or `reconfigure(line_buffering=True)`) would make logs trustworthy.
5. **Document `NCCL_P2P_DISABLE=1`** as a known requirement for multi-GPU training on
   workstations without NVLink — this affects any AMD/IOMMU box, which is a lot of them.
</details>

**Comment by McHaillet** [comment]
Thanks for reporting @hamid13r ! Unclear to me why this happens, but the suggested fix is easy at least. Someone else had a similar problem with the P2P not working on a cluster and then we also came to this fix, so it's in the docs now at least.
